src/gui/widgets/combobox.cpp

```diff
@@ -0,0 +1,491 @@
+	variables.add("text_font_height", wfl::variant(text_height_));
+
+	text_x_offset_ = conf->text_x_offset(variables);
+	text_y_offset_ = conf->text_y_offset(variables);
+
+	// Since this variable doesn't change set it here instead of in
+	// update_canvas().
+	for(auto & tmp : get_canvases())
+	{
+		tmp.set_variable("text_font_height", wfl::variant(text_height_));
+	}
+
+	// Force an update of the canvas since now text_font_height is known.
+	update_canvas();
+}
+
+void combobox::handle_key_clear_line(SDL_Keymod /*modifier*/, bool& handled)
+{
+	handled = true;
+	set_value("");
+}
+
+void combobox::handle_key_up_arrow(SDL_Keymod /*modifier*/, bool& handled)
+{
+	DBG_GUI_E << LOG_SCOPE_HEADER;
+	handled = true;
+	if (selected_ > 1) {
+		set_selected(selected_ - 1, true);
+	}
+}
+
+void combobox::handle_key_down_arrow(SDL_Keymod /*modifier*/, bool& handled)
+{
+	DBG_GUI_E << LOG_SCOPE_HEADER;
+	handled = true;
+	if (selected_ < values_.size()-1) {
+		set_selected(selected_ + 1, true);
+	}
+}
+
+void combobox::set_values(const std::vector<::config>& values, unsigned selected)
+{
+	assert(selected < values.size());
+	assert(selected_ < values_.size());
+
+	if(values[selected]["label"] != values_[selected_]["label"]) {
+		queue_redraw();
+	}
+
+	values_ = values;
+	selected_ = selected;
+
+//	set_label(values_[selected_]["label"]);
+	text_box_base::set_value(values_[selected_]["label"]);
+}
+
+void combobox::set_selected(unsigned selected, bool fire_event)
+{
+	assert(selected < values_.size());
+	assert(selected_ < values_.size());
+
+	if(selected != selected_) {
+		queue_redraw();
+	}
+
+	selected_ = selected;
+
+//	set_label(values_[selected_]["label"]);
+	text_box_base::set_value(values_[selected_]["label"]);
+	if (fire_event) {
+		fire(event::NOTIFY_MODIFIED, *this, nullptr);
+	}
+}
+
+void combobox::update_mouse_cursor()
+{
+	unsigned x = get_x() + this->get_size().x;
+	unsigned mx = get_mouse_position().x;
+
+	if ((mx <= x) && (mx >= x-ICON_SIZE)) {
+		cursor::set(cursor::NORMAL);
+	} else {
+		cursor::set(cursor::IBEAM);
+	}
+}
+
+void combobox::signal_handler_mouse_enter(const event::ui_event /*event*/,
+											   bool& /*handled*/)
+{
+	update_mouse_cursor();
+}
+
+void combobox::signal_handler_mouse_motion(const event::ui_event event,
+											bool& handled,
+											const point& coordinate)
+{
+	DBG_GUI_E << get_control_type() << "[" << id() << "]: " << event << ".";
+
+	if(dragging_) {
+		handle_mouse_selection(coordinate, false);
+	} else {
+		update_mouse_cursor();
+	}
+
+	handled = true;
+}
+
+void combobox::signal_handler_left_button_down(const event::ui_event event,
+												bool& handled)
+{
+	DBG_GUI_E << LOG_HEADER << ' ' << event << ".";
+
+	/* get_x() is the left border
+	 * this->get_size().x is the size of this widget
+	 * so get_x() + this->get_size().x is the right border
+	 * ICON_SIZE is the size of the icon. */
+
+	unsigned x = get_x() + this->get_size().x;
```

Comment:
I think calling this `right_border` or so would be more readable.

Reply:
Changed.